🎚️ feat: Custom Parameters

nhtruong opened the pull request:
This pull request introduces support for custom parameters in configuration endpoints, refactors existing code for simplification, and updates the client-side components to handle dynamic settings more effectively.

This PR is accompanied by this [Documentation update](https://github.com/LibreChat-AI/librechat.ai/pull/297).

## Summary
### New Feature: Picking A Default Parameters Set
By default, when you specify a custom endpoint in `librechat.yaml` config file, it will use the default parameters of the OpenAI API. However, you can override these defaults by specifying the `customParams.defaultParamsEndpoint` field within the definition of your custom endpoint. For example, to use Google parameters for your custom endpoint:

```yaml
endpoints:
  custom:
    - name: 'Google Gemini'
      apiKey: ...
      baseURL: ...
      customParams:
        defaultParamsEndpoint: 'google'
 ```
Your “Google Gemini” endpoint will now display parameters for Google API when you create a new agent or preset.

### New Feature: Overriding Parameter Definitions

On top of that, you can also fine tune the parameters provided for your custom endpoint. For example, if you want to limit the range of the `temperature` parameter to a range between `0.0` and `0.7` with step of `0.1` and default of `0.5`, your `librechat.yaml` file would look like this:

```yaml
endpoints:
  custom:
    - name: 'Google Gemini'
      apiKey: ...
      baseURL: ...
      customParams:
        defaultParamsEndpoint: 'google'
        paramDefinitions:
          - key: temperature
            range:
              min: 0
              max: 0.7
              step: 0.1
            default: 0.5
```

![Screenshot 2025-05-12 at 4 32 07 PM](https://github.com/user-attachments/assets/3284adac-f826-456c-b484-486ac4f5b030)

### Refactor:
- moved `Parameters/settings.ts` into `data-provider` so that both frontend and backend code can use it.
- simplified `getCustomConfig` function

## Change Type

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Checklist

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented in any complex areas of my code
- [x] I have made pertinent documentation changes
- [x] My changes do not introduce new warnings
- [x] I have written tests demonstrating that my changes are effective or that my feature works
- [x] Local unit tests pass with my changes
- [x] A pull request for updating the documentation has been submitted.

nhtruong committed:
#

nhtruong committed:
- refactor: simplified getCustomConfig func

nhtruong committed:
#

nhtruong committed:
- feature: persist values for parameters with optionType of custom

nhtruong committed:
#

nhtruong committed:
- refactor: moved `Parameters/settings.ts` into `data-provider` so that both frontend and backend code can use it.

nhtruong committed:
- feature: loadCustomConfig can now parse and validate customParams property for `endpoints.custom` in `librechat.yaml`

danny-avila commented:
Thanks @nhtruong will test soon.

Just some ESLint and unused localization string issues for now.

nhtruong commented:
Thanks, @danny-avila. I'll take care of those.

I also have a question for you. When constructing the overridden parameters, I have to make sure that `optionType` is always `custom` as seen below:
https://github.com/danny-avila/LibreChat/blob/79fe2c6a5d5e36689b71e3a9cef2ddbf7fbc665d/api/server/services/Config/loadCustomConfig.js#L160
This is because `validateSettingDefinitions` 

https://github.com/danny-avila/LibreChat/blob/79fe2c6a5d5e36689b71e3a9cef2ddbf7fbc665d/api/server/services/Config/loadCustomConfig.js#L164
would throw an error otherwise:
```
"Setting temperature with optionType \"model\" must match the type defined in tConversationSchema."
```
This led to changes to the React components in `SidePanel/Parameters/` folder to persist parameter values of parameters with optionType of `custom`. I'm not sure if this is the right approach or if it would cause any side effects.

---
 I've investigated why `validateSettingDefinitions` would throw errors when validating `optionType: model` parameters when it reaches this part:
https://github.com/danny-avila/LibreChat/blob/7b23131cb9926e68011cad36c389285c04a7126a/packages/data-provider/src/generate.ts#L456-L464
I think it's a bug where `zodType` is `ZodOptional` while it should have been ZodNumber for `temperature` parameter:
https://github.com/danny-avila/LibreChat/blob/66093b1eb38462b74d187a89a8f4d0075f40d2c4/packages/data-provider/src/schemas.ts#L578
This is because ZodNumber, the value we're looking for, is in the nested typeName, not the root-level typeName:
```json
# conversationSchema._def
{
  "innerType": {
    "_def": {
      "checks": [],
      "typeName": "ZodNumber",
      "coerce": false
    },
    "~standard": {
      "version": 1,
      "vendor": "zod"
    }
  },
  "typeName": "ZodOptional"
}
```

---

If `optionType: custom` is not a problem, then I will create a seperate PR to fix said bug. If these overridden params really need to be of the same `optionType` as the default params, then I'll update this PR accordingly with the bug fix.

mawburn commented on client/src/components/SidePanel/Agents/ModelPanel.tsx:
```diff
@@ -1,16 +1,16 @@
 import { getEndpointField, cn } from '~/utils';
 import { useLocalize } from '~/hooks';
 import { Panel } from '~/common';
+import _ from 'lodash';
```
If you did decide to keep lodash (although, I don't think you should), we should import them in a way that doesn't significantly impact the bundle size. I think doing this will result in >70kb and do something like this:

```suggestion
import keys from 'lodash/keys';
```

This is less important on the backend, but it's still an optimization that's worth it.

mawburn commented on client/src/components/SidePanel/Parameters/DynamicSwitch.tsx:
```diff
@@ -35,20 +33,11 @@ function DynamicSwitch({
-      return inputValue;
-    }
-
     return conversation?.[settingKey] ?? defaultValue;
-  }, [conversation, defaultValue, optionType, settingKey, inputValue]);
+  }, [conversation, defaultValue, settingKey]);
```
We could probably make this cleaner by doing: 

```typescript
const selectedValue = useMemo(() => 
  conversation?.[settingKey] ?? defaultValue, 
  [conversation?.[settingKey], defaultValue]
);
```

I might be wrong, I didn't do this in the editor but `conversation?.[settingKey]` will also only cause a rerender when the specific value changes instead of the whole object.

nhtruong replied:
In this case, I don't think useMemo is actually counterproductive as the overhead memoization and of tracking changes in dependencies costs a lot more than nullish coalescing operations. So we should change this to:

```ts
const selectedValue = conversation?.[settingKey] ?? defaultValue
```

mawburn commented on client/src/components/SidePanel/Parameters/DynamicTags.tsx:
```diff
@@ -54,18 +48,9 @@ function DynamicTags({
-
-    return conversation[settingKey];
-  }, [conversation, defaultValue, optionType, settingKey, tags]);
+  const currentTags = useMemo(() => {
+    return conversation?.[settingKey] ?? defaultValue ?? [];
+  }, [conversation, defaultValue, settingKey]);
```
Same thing here. 

```typescript
const currentTags = useMemo(() =>
   conversation?.[settingKey] ?? defaultValue ?? [], 
   [conversation?.[settingKey], defaultValue]
);
```

nhtruong replied:
Same response as above.

mawburn reviewed (commented):
I just did a quick review with a few changes. I really think we should get rid of lodash as it's not needed.

nhtruong committed:
# fixed linter

nhtruong committed:
# removed .strict() in config.ts

nhtruong committed:
change: added packages/data-provider/src to SOURCE_DIRS for i18n check

nhtruong committed:
# removed unnecessary lodash imports

nhtruong commented:
@mawburn addressed your concerns over lodash with [this commit](https://github.com/danny-avila/LibreChat/pull/7342/commits/1411da9b791b908735c23c73790c99cb7fa65aab).

sbruel commented on client/src/components/SidePanel/Parameters/Panel.tsx:
```diff
@@ -29,13 +37,17 @@ export default function Parameters() {
-    const [combinedKey, endpointKey] = getSettingsKeys(
-      endpointType ?? conversation?.endpoint ?? '',
-      conversation?.model ?? '',
+  const parameters = useMemo((): SettingDefinition[] => {
```
Can we return the overriddenParams from the API and keep the frontend agnostic of it?

sbruel replied:
I haven't tracked the full path, but I assume the `endpointsConfig`  comes from the `endpointController` in the backend. 
Could we transform the object there to be the final shape the frontend needs it so the frontend doesn't need to know about `defaultParams` or `overriddenParams` , just fully materialized endpointsConfig

nhtruong replied:
Good point. This same problem exist before this PR tho. The frontend still needs to calculate the list of params based on the endpoint being returned by the API. It's outside the scope of this feature. I'll create an issue for this.

nhtruong commented on api/server/services/Config/loadCustomConfig.js:
```diff
@@ -117,4 +129,52 @@ https://www.librechat.ai/docs/configuration/stt_tts`);
+  customParams.paramDefinitions = customParams.paramDefinitions || [];
+
+  // Checks if `defaultParamsEndpoint` is a key in `paramSettings`.
+  const validEndpoints = new Set([
```
But this is the only place it's being used. What's the goal of moving it out of the function?

nhtruong replied:
> it will potentially be called for every endpoint no?
Actually no. `validEndpoints` is just a list of endpoint names used for validation purpose when the librechat.yaml file is loaded up.

nhtruong commented on packages/data-provider/src/config.ts:
```diff
@@ -278,6 +278,12 @@ export const endpointSchema = baseEndpointSchema.merge(
+    customParams: z
+      .object({
+        defaultParamsEndpoint: z.string().default('custom'),
+        paramDefinitions: z.array(z.record(z.any())).optional(),
```
The validation required for a parameter definition is beyond what zod can provide. It involves a lot of if/else conditions. Say if `type: number` then `default:` must be of type number as well. That's why it has its own custom validation logic invoked here: https://github.com/danny-avila/LibreChat/pull/7342/files#diff-ee88da5c94df12ca91ff12251f99a4161e048c0ecefba951d8b6776d06a2380aR172

nhtruong commented on packages/data-provider/src/types.ts:
```diff
@@ -268,6 +268,10 @@ export type TConfig = {
   capabilities?: string[];
+  customParams?: {
+    defaultParamsEndpoint?: string;
+    paramDefinitions?: Array<object>;
```
This one is tricky because it depends on what set of definition is being pulled out of paramSettings:
https://github.com/danny-avila/LibreChat/blob/f18c031707c040b21ccb2b666b85b4e4632cf60b/packages/data-provider/src/parameterSettings.ts#L657-L670
or agentParamSettings:
https://github.com/danny-avila/LibreChat/blob/f18c031707c040b21ccb2b666b85b4e4632cf60b/packages/data-provider/src/parameterSettings.ts#L719-L726

nhtruong committed:
# addressed PR comments
# fixed lint for updated files

mawburn commented on .github/workflows/i18n-unused-keys.yml:
```diff
@@ -22,7 +22,7 @@ jobs:
           # Define paths
           I18N_FILE="client/src/locales/en/translation.json"
-          SOURCE_DIRS=("client/src" "api")
+          SOURCE_DIRS=("client/src" "api" "packages/data-provider/src")
```
Why is this needed? data-provider doesn't have any i18n keys does it?

nhtruong replied:
Because [parameterSettings](https://github.com/danny-avila/LibreChat/pull/7342/files/a36f58467dd217fa832e81920d55d07d5923e6fd#diff-f6a5533d3123073da033a3462eee314c8af6118b55d78e22f26039dc95e70514) has been moved from frontend to data-provider so that the backend can use it to generate overriden params and validate their definitions.

This really should be the responsibility of the backend, and the frontend should only be responsible for displaying a list of params given to it (i.e. the param selection/parsing logic should be handled on the backend). It's related to this [comment](https://github.com/danny-avila/LibreChat/pull/7342#discussion_r2089279647) by @sbruel. I'm creating a follow-up issue to handle this isolation-of-concerns issue.

mawburn commented on api/server/services/Config/getCustomConfig.js:
```diff
@@ -10,17 +10,7 @@ const getLogStores = require('~/cache/getLogStores');
-  }
-
-  return customConfig;
+  return (await cache.get(CacheKeys.CUSTOM_CONFIG)) || (await loadCustomConfig());
```
This changes at the logic a little bit and doesn't return null from the function anymore. Are you sure that's not needed?

nhtruong replied:
`await loadCustomConfig()` already returns null in some cases and a config object in others. So `(await loadCustomConfig()) || null` is not necessary.

mawburn commented on client/src/components/SidePanel/Agents/ModelPanel.tsx:
```diff
@@ -1,16 +1,21 @@
 import { getEndpointField, cn } from '~/utils';
 import { useLocalize } from '~/hooks';
 import { Panel } from '~/common';
+import { keyBy } from 'lodash';
```
I know lodash has tree shaking now, but the last time I checked it isn't perfect. But if you import it like this you don't have to worry about it. 

```suggestion
import keyBy from 'lodash/keyBy';
```

nhtruong committed:
# better import for lodash (w/o relying on tree-shaking)

nhtruong commented:
@danny-avila this has stabilized. All concerns from @mawburn and @sbruel have been addressed. Lemme know if there's anything else I should update.

danny-avila commented:
Thank you @nhtruong I'll be testing this today and over the weekend, hope to merge by Monday.
